### PR TITLE
allow auto running of command returned from chat response

### PR DESCRIPTION
**Pull request by meganrogge**
wip #209906

To do:
- Consider UX
- Verify safe command list

### CODE REVIEW

File: src/vs/workbench/contrib/chat/browser/chatCommandExecutionService.ts [hunk 1]
```diff
@@ -0,0 +1,911 @@
...
+		super();
+	}
+	async acceptResponse(response: IChatResponseViewModel | string | undefined): Promise<void> {
+		if (!this._configurationService.getValue('github.copilot.experimental.runBestCommandMatch')) {
+			return;
+		}
+		if (typeof response === 'string') {
+			return;
+		}
+		const button = response?.renderData?.renderedParts.find(part => isCommandButtonRenderData(part));
```
Comment (roblourens) [inline src/vs/workbench/contrib/chat/browser/chatCommandExecutionService.ts:34]: You shouldn't be looking inside renderData, it's really just private data for the list renderer, you can just look at the response data itself (the `response`) propery.

File: src/vs/workbench/contrib/chat/browser/chatCommandExecutionService.ts [hunk 2]
```diff
@@ -0,0 +1,911 @@
...
+			if (confirmation.checkboxChecked === true) {
+				this._storageService.store(GRANTED_RUN_COMMAND_PERMISSION_KEY, true, StorageScope.WORKSPACE, StorageTarget.USER);
+			}
+		}
+
+		this._commandService.executeCommand(button.command.id);
+	}
+}
+
+const commandsToRun = [
```
Comment (roblourens) [inline src/vs/workbench/contrib/chat/browser/chatCommandExecutionService.ts:65]: Might as well make this a Set when it's this large

File: src/vs/workbench/contrib/chat/browser/chatCommandExecutionService.ts [hunk 3]
```diff
@@ -0,0 +1,911 @@
...
+				return;
+			}
+
+			// Check for confirmation checkbox
+			if (confirmation.checkboxChecked === true) {
+				this._storageService.store(GRANTED_RUN_COMMAND_PERMISSION_KEY, true, StorageScope.WORKSPACE, StorageTarget.USER);
+			}
+		}
+
+		this._commandService.executeCommand(button.command.id);
```
Comment (roblourens) [inline src/vs/workbench/contrib/chat/browser/chatCommandExecutionService.ts:61]: Should this also use the args from the button?

File: src/vs/workbench/contrib/chat/browser/chatCommandExecutionService.ts [hunk 4]
```diff
@@ -0,0 +1,911 @@
...
+	}
+	async acceptResponse(response: IChatResponseViewModel | string | undefined): Promise<void> {
+		if (!this._configurationService.getValue('github.copilot.experimental.runBestCommandMatch')) {
+			return;
+		}
+		if (typeof response === 'string') {
+			return;
+		}
+		const button = response?.renderData?.renderedParts.find(part => isCommandButtonRenderData(part));
+		if (!button || !('command' in button) || !commandsToRun.includes(button.command.id)) {
```
Comment (roblourens) [inline src/vs/workbench/contrib/chat/browser/chatCommandExecutionService.ts:35]: I think this should only be enabled for the `@vscode` participant or participants that opt-in to this, which they could do by setting another flag when they stream back the button. It makes sense when you are asking it to recommend a command, and might not make sense at all in a different scenario.

File: src/vs/workbench/contrib/chat/browser/chatCommandExecutionService.ts [hunk 5]
```diff
@@ -0,0 +1,911 @@
...
+			return;
+		}
+		const grantedPermission = this._storageService.getBoolean(GRANTED_RUN_COMMAND_PERMISSION_KEY, StorageScope.WORKSPACE, false);
+		if (!grantedPermission) {
+			const message = localize('confirmAutomaticRunChatResponseCommand', "Are you sure you want to automatically run the command from the chat response?");
```
Comment (roblourens) [inline src/vs/workbench/contrib/chat/browser/chatCommandExecutionService.ts:40]: Seems off to say "Are you sure" when I didn't take any action to make this appear. How about "Would you like to automatically run the command in the chat response" and the checkbox can say "Always automatically run commands" or "Don't ask me again"

File: src/vs/workbench/contrib/chat/browser/chatCommandExecutionService.ts (hunk 1 above)
Comment (roblourens) [reply src/vs/workbench/contrib/chat/browser/chatCommandExecutionService.ts:34]: There's probably a race here actually

File: src/vs/workbench/contrib/chat/browser/chatCommandExecutionService.ts (hunk 5 above)
Comment (meganrogge) [reply src/vs/workbench/contrib/chat/browser/chatCommandExecutionService.ts]: I think asking to run the command with a checkbox to always allow this makes sense. I will add a button to never ask again.

Comment (meganrogge) [reply src/vs/workbench/contrib/chat/browser/chatCommandExecutionService.ts]: Actually, i don't think we need the checkbox. We can just ask this question.

File: src/vs/workbench/contrib/chat/browser/chatCommandExecutionService.ts (hunk 4 above)
Comment (meganrogge) [reply src/vs/workbench/contrib/chat/browser/chatCommandExecutionService.ts]: I have hardcoded it for the `vscode` agent for now. Do you mean adding a property to `IResponse` like `allowAutomaticRunCommands`?

**Pull request closed without merging**